### ISSUE TITLE
Task: Do not allow for B to be derived from a function call or other hard-to-reason about structures

### ISSUE THREAD

**Issue by thomasrolinger** [thomasrolinger/chapel#4]
Something like `ref B = someFunc()` is a problem because we don't know what's going in within `someFunc()`. Specifically, we cannot reason about whether `B` will be different each time `someFunc()` is called. A crucial component in our optimization is knowing when to re-run the inspector, which partially hinges on when `B` is changed.

We allow for `B` to be defined within the loop body if it is a `ref` to something outside of the loop. We just need to add an extra condition here that `B` cannot be the return/result of a function call.

However, it can get a bit tricky in that if `B` is defined "sufficiently" far outside of the loop, then it really isn't a problem for it to be derived from a function call. You can imagine some sort of initialization routine. However, as I am typing this I am wondering whether we would even make it that far. I noticed in the original PageRank code that we had `var A = readGraph()` and that caused an issue because we couldn't find `A`'s domain symbol. I assume if we had `ref B = someFunc()` we would arrive at the same problem (though in this case, it isn't a problem for us).

Another case to consider is something like `ref B = arrays[n]` where `arrays` is an array of arrays and `n` is some variable that is NOT yielded by the loop(s). That is a problem for similar reasons, as we cannot reason about "which" `B` we will get whenever the loop runs. From what I can tell in my in-code comments, the call to `isSymbolDefinedInBlock` would return `B` in this case and not find anything wrong with it. So we may need to add a special check here, but to be honest, it is more of a production concern right now.

So to summarize:

* we may not need to do anything for the case when `B` is the result of a function call, but we should verify this
* we may need additional checks for `ref B = arrays[n]` but it isn't a top concern right now

**Comment by thomasrolinger** [comment]
Not going to worry about this right now. We no longer allow for `B` to be defined in the loop, so this is not a big concern. I think there will may be an edge case or two that we would care about eventually, but nothing that is present in our current use-cases.
